pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,318 @@
+                # Given that we only used Z as generator, we need to apply some gates before and after the generator.
+                if p_idx == 0:
+                    op_temp = ops_to_trainable_op.pop(-1)
+                    ops_after_trainable_op = [op_temp] + ops_after_trainable_op
+                elif p_idx == 1:
+                    ops_to_add_before = [
+                        qml.RZ(-trainable_op.data[2], wires=trainable_op.wires),
+                        qml.RX(qml.numpy.pi / 2, wires=trainable_op.wires),
+                    ]
+                    ops_to_trainable_op.extend(ops_to_add_before)
+
+                    ops_to_add_after = [
+                        qml.RX(-qml.numpy.pi / 2, wires=trainable_op.wires),
+                        qml.RZ(trainable_op.data[2], wires=trainable_op.wires),
+                    ]
+                    ops_after_trainable_op = ops_to_add_after + ops_after_trainable_op
+
+            ctrl_gen = [qml.ctrl(gen, control=aux_wire)]
+            hadamard = [qml.Hadamard(wires=aux_wire)]
+            ops = ops_to_trainable_op + hadamard + ctrl_gen + hadamard + ops_after_trainable_op
+
+            measurements = []
+            # Add the Y measurement on the aux qubit
+            for m in tape.measurements:
+
+                if isinstance(m.obs, qml.operation.Tensor):
+                    obs_new = m.obs.obs.copy()
+                else:
+                    if m.obs:
+                        obs_new = [m.obs]
+                    else:
+                        obs_new = [qml.PauliZ(wires=i) for i in m.wires]
+
+                obs_new.append(qml.PauliY(wires=aux_wire))
+                obs_new = qml.operation.Tensor(*obs_new)
+
+                if isinstance(m, qml.measurements.ExpectationMP):
+                    measurements.append(qml.expval(op=obs_new))
+                else:
+                    measurements.append(qml.probs(op=obs_new))
+
+            new_tape = qml.tape.QuantumTape(ops=ops, measurements=measurements)
+
+            def stop_at(obj):
+                return ~qml.operation.is_measurement(obj)
+
+            # Expand measurements only
+            new_tape = qml.tape.tape.expand_tape(
+                new_tape, stop_at=stop_at, expand_measurements=True
+            )
+            num_tape += 1
+
+            g_tapes.append(new_tape)
+
+        gradient_data.append(num_tape)
+
+    def processing_fn(results):  # pylint: disable=too-many-branches
+        # TODO interface compatibility
+        multi_measurements = len(tape.measurements) > 1
+        multi_params = len(tape.trainable_params) > 1
+
+        grads = []
+        final_res = []
+        for coeff, res in zip(coeffs, results):
+            if isinstance(res, tuple):
+                final_res.append([2 * coeff * r for r in res])
+            else:
+                final_res.append(2 * coeff * res)
+
+        # Post process for probs
+        if measurements_probs:
+            projector = np.array([1, -1])
```

Comment:
might this trigger a warning from `get_interface` in the `qml.math.tensordot` below if `res`/`res_reshaped` is from some other interface?